### PR TITLE
Pkcs11 login

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: host/xtest/pkcs11_1000.c [hunk 1]
```diff
@@ -419,3 +419,158 @@ static void xtest_pkcs11_test_1002(ADBG_Case_t *c)
...
+	if (rv != CKR_OK)
+		goto out;
+
+out:
+	rv = close_lib();
+	ADBG_EXPECT_CK_OK(c, rv);
+}
+
+ADBG_CASE_DEFINE(pkcs11, 1003, xtest_pkcs11_test_1003,
+		"PKCS11: Login to PKCS#11 token");
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:847]: indent with one more space char.

File: host/xtest/pkcs11_1000.c [hunk 2]
```diff
@@ -419,3 +419,158 @@ static void xtest_pkcs11_test_1002(ADBG_Case_t *c)
...
+ * Helpers for tests where we must log into the token.
+ * These define the genuine PINs and label to be used with the test token.
+ */
+static CK_UTF8CHAR test_token_so_pin[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8 , 9, 10,};
+static CK_UTF8CHAR test_token_label[] = "PKCS11 TA test token";
+
+static CK_RV init_test_token(CK_SLOT_ID slot)
+{
+	return C_InitToken(slot,
+			   (CK_UTF8CHAR_PTR)test_token_so_pin,
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:436]: cast not needed here and below.

File: host/xtest/pkcs11_1000.c [hunk 3]
```diff
@@ -425,6 +425,9 @@ ADBG_CASE_DEFINE(pkcs11, 1002, xtest_pkcs11_test_1002,
  * These define the genuine PINs and label to be used with the test token.
  */
 static CK_UTF8CHAR test_token_so_pin[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8 , 9, 10,};
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:427]: add a space before `}`

File: host/xtest/pkcs11_1000.c [hunk 4]
```diff
@@ -626,6 +626,87 @@ static CK_RV test_uninitialized_token(ADBG_Case_t *c, CK_SLOT_ID slot)
...
+
+	rv = C_Login(session, CKU_SO, test_token_so_pin,
+		     sizeof(test_token_so_pin));
+	ADBG_EXPECT_CK_RESULT(c, CKR_USER_ALREADY_LOGGED_IN, rv);
+
+	rv = C_Logout(session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	/* Login user then so and reverse */
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:673]: s/so/security officer/  or  s/so/SO/

File: host/xtest/pkcs11_1000.c [hunk 5]
```diff
@@ -639,6 +720,13 @@ static void xtest_pkcs11_test_1003(ADBG_Case_t *c)
 	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_InitPIN) ||
 	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_SetPIN))
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:790]: remove the test above.

File: host/xtest/pkcs11_1000.c [hunk 6]
```diff
@@ -662,6 +750,12 @@ static void xtest_pkcs11_test_1003(ADBG_Case_t *c)
 	if (rv != CKR_OK)
 		goto out;
 
+	rv = test_login_logout(c, slot);
+	if (rv != CKR_OK)
+		goto out;
+
+
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:829]: some space for further tests? :)

File: host/xtest/pkcs11_1000.c [hunk 7]
```diff
@@ -707,6 +707,75 @@ static CK_RV test_login_logout(ADBG_Case_t *c, CK_SLOT_ID slot)
...
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out_session;
+	}
+
+	rv = C_SetPIN(session, old_pin, old_pin_sz, some_pin, sizeof(some_pin));
+	if (!ADBG_EXPECT_CK_OK(c, rv)) {
+		if (ut == user_type)
+			goto out_logout;
+		else
+			goto out_session;
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:748]: Failing `C_SetPIN()` is not expected to log out the currently logged user. This should `goto out_logout;` in both cases.

File: host/xtest/pkcs11_1000.c [hunk 8]
```diff
@@ -707,6 +707,75 @@ static CK_RV test_login_logout(ADBG_Case_t *c, CK_SLOT_ID slot)
...
+	rv2 = C_CloseSession(session);
+	if (!ADBG_EXPECT_CK_OK(c, rv2) && !rv)
+		rv = rv2;
+out:
+	Do_ADBG_EndSubCase(c, "Test C_SetPIN() user_type %lu", user_type);
+
+	return rv;
+}
+
+
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:778]: remove 1 empty line

File: host/xtest/pkcs11_1000.c [hunk 9]
```diff
@@ -754,8 +823,21 @@ static void xtest_pkcs11_test_1003(ADBG_Case_t *c)
...
+	if (rv != CKR_OK)
+		goto out;
 
+	/*
+	 * CKU_CONTEXT_SPECIFIC is anything not CKU_USER or CKU_SO in order
+	 * to skip the initial login.
+	 */
+	rv = test_set_pin(c, slot, CKU_CONTEXT_SPECIFIC);
+	if (rv != CKR_OK)
+		goto out;
```
Comment (etienne-lms) [inline host/xtest/pkcs11_1000.c:840]: add an empty line below. Or maybe remove lines 839 and 840.

File: host/xtest/pkcs11_1000.c (hunk 1 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:847]: I'll fix

File: host/xtest/pkcs11_1000.c (hunk 2 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:436]: I'll fix

File: host/xtest/pkcs11_1000.c (hunk 3 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:427]: OK

File: host/xtest/pkcs11_1000.c (hunk 5 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:790]: OK

File: host/xtest/pkcs11_1000.c (hunk 6 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:829]: Yeah! :-)

File: host/xtest/pkcs11_1000.c (hunk 7 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:748]: In the case where `user_type` isn't `CKU_SO` or `CKU_USER` we're not logged in, that's why we're skipping the logout.

File: host/xtest/pkcs11_1000.c (hunk 9 above)
Comment (jenswi-linaro) [reply host/xtest/pkcs11_1000.c:840]: I'll remove this redundant test.

**Comment by jenswi-linaro** [comment]
Update

**Comment by jenswi-linaro** [comment]
Squashed, rebased and tag applied.

**Comment by jenswi-linaro** [comment]
Update

**Pull request merged**